pyvrp/cpp/repair/greedy_repair.cpp

```diff
@@ -0,0 +1,152 @@
+#include "greedy_repair.h"
+
+#include "TimeWindowSegment.h"
+#include "search/Route.h"
+
+#include <cassert>
+
+using pyvrp::CostEvaluator;
+using pyvrp::DynamicBitset;
+using pyvrp::ProblemData;
+using pyvrp::Solution;
+using pyvrp::search::n;
+using pyvrp::search::p;
+using pyvrp::search::Route;
+
+using Clients = std::vector<Route::Node>;
+using Routes = std::vector<Route>;
+using TWS = pyvrp::TimeWindowSegment;
+
+namespace
+{
+// Populate the given clients and routes vectors with routes from the solution.
+void setupRoutes(Clients &clients,
+                 Routes &routes,
+                 Solution const &sol,
+                 ProblemData const &data)
+{
+    assert(clients.empty() && routes.empty());
+
+    // Doing this avoids re-allocations, which would break the pointer structure
+    // that Route and Route::Node use.
+    clients.reserve(data.numClients() + 1);
+    routes.reserve(sol.numRoutes());
+
+    for (size_t client = 0; client <= data.numClients(); ++client)
+        clients.emplace_back(client);
+
+    size_t idx = 0;
+    for (auto const &solRoute : sol.getRoutes())
+    {
+        auto &route = routes.emplace_back(data, idx++, solRoute.vehicleType());
+        for (auto const client : solRoute)
+            route.push_back(&clients[client]);
+
+        route.update();
+    }
+}
+
+// Turns the given routes into a solution.
+Solution exportRoutes(ProblemData const &data, Routes const &routes)
+{
+    std::vector<Solution::Route> solRoutes;
+    solRoutes.reserve(routes.size());
+
+    for (auto const &route : routes)
+    {
+        std::vector<size_t> visits;
+        visits.reserve(route.size());
+
+        for (auto *node : route)
+            visits.push_back(node->client());
+
+        solRoutes.emplace_back(data, visits, route.vehicleType());
+    }
+
+    return {data, solRoutes};
+}
+
+// Evaluates the move of inserting U after V.
+pyvrp::Cost evaluate(Route::Node *U,
```

Comment:
This has overlap with LS::maybeInsert, and obviously also with (1, 0)-exchange. I'll make an issue to track that we could perhaps deduplicate this at some point.

Reply:
#341